# HiSST schemes: databases and R scripts
HiSST scheme databases and R-scripts used for opportunistic pathogens pure culture and environmental DNA survey, using the Illumina next-generation sequencing (NGS) method.

### Bioinformatical pipeline for HiSST analysis:
Originate from dada2 pipeline (https://benjjneb.github.io/dada2/index.html) and was adapted for HiSST schemes analysis, including additional steps to enhance the accuracy of taxonomic assignment. 
The entire process can be executed with the R script [Script_RUN_FunHiSSTDada2.R](https://github.com/LaboPC/HiSST-schemes_TB_2023/blob/main/Script_RUN_FunHiSSTDada2.R), using a unified function called [FunHiSSTDada2.R](https://github.com/LaboPC/HiSST-schemes_TB_2023/blob/main/FunHiSSTDada2.R).

# HiSST scheme for _Serratia marcescens_ :

The databases for each locus are available in the [Serratia_marcescens](https://github.com/LaboPC/HiSST-schemes_TB_2023/tree/main/Serratia_marcescens) directory file.

- R-scripts used to develop, test and validate the HiSST scheme for _Serratia marcescens_ are available at https://github.com/TBourd/R_scripts_for_HiSST_scheme

- HiSST application example: R-scripts and data used for _Serratia marcescens_ colonizations survey, using the HiSST scheme, are available at https://github.com/TBourd/R_scripts_HiSST_SM-colonizations (see article: Bourdin et al., 2023)

###### If you use this HiSST scheme in your research, please cite :
> Bourdin T, Monnier A, Benoit MÈ, Bédard E, Prévost M, Quach C, Déziel E, Constant P. A High-Throughput Short Sequence Typing Scheme for Serratia marcescens Pure Culture and Environmental DNA. Appl Environ Microbiol. 2021. DOI: https://doi.org/10.1128/AEM.01399-21


# HiSST scheme for _Pseudomonas aeruginosa_ :

The databases for each locus are available in the [Pseudomonas_aeruginosa](https://github.com/LaboPC/HiSST-schemes_TB_2023/tree/main/Pseudomonas_aeruginosa) directory file.

> Description will be available soon


# HiSST scheme for _Stenotrophomonas maltophilia_ :

The databases for each locus are available in the [Stenotrophomonas_maltophilia](https://github.com/LaboPC/HiSST-schemes_TB_2023/tree/main/Stenotrophomonas_maltophilia) directory file.

> Description will be available soon

 _______________________________________________________

# Publications

 
- Bourdin T, Monnier A, Benoit MÈ, Bédard E, Prévost M, Quach C, Déziel E, Constant P. 2021. A High-Throughput Short Sequence Typing Scheme for Serratia marcescens Pure Culture and Environmental DNA. Appl Environ Microbiol. DOI: https://doi.org/10.1128/AEM.01399-21

- Bourdin T, Benoit M-È, Monnier A, Bédard E, Prévost M, Charron D, Audy N, Gravel S, Sicard M, Quach C, Déziel E, Constant P. 2023. Serratia marcescens Colonization in a Neonatal Intensive Care Unit Has Multiple Sources, with Sink Drains as a Major Reservoir. Applied and Environmental Microbiology 89:e00105-23. DOI: https://doi.org/10.1128/aem.00105-23

- Bourdin T, Benoit M-È, Bédard E, Prévost M, Quach C, Déziel E, Constant P. _Submitted for publication_. High-Throughput Short Sequence Typing Schemes for Pseudomonas aeruginosa and Stenotrophomonas maltophilia pure culture and environmental DNA.
